lib/pbench/server/api/resources/query_apis/datasets_publish.py

```diff
@@ -0,0 +1,210 @@
+        # pyesbulk here; but I don't want to duplicate the setup in indexer.py,
+        # it's not well generalized, and I don't want to add that refactoring to
+        # this PR.
+
+        cmds = []
+        self.logger.info("MAP: {!r}", map)
+        for index, ids in map.items():
+            for id in ids:
+                x = (
+                    f'{{"update": {{"_index": "{index}", "_id": "{id}"}}}}\n'
+                    f'{{"doc": {{"authorization": {{"access": "{access}"}}}}}}'
+                )
+                cmds.append(x)
+        data = "\n".join(cmds) + "\n"
+        return {
+            "path": "/_bulk",
+            "kwargs": {
+                "data": data,
+                "headers": {"Content-Type": "application/x-ndjson"},
+                "params": {"refresh": "true"},
+            },
+        }
+
+    def postprocess(
+        self, es_json: Dict[AnyStr, Any], context: Dataset
+    ) -> Dict[AnyStr, Any]:
+        """
+        A bulk update succeeds if any of the individual updates succeeds, and
+        PUBLISH can affect hundreds of documents across many Elasticsearch
+        indices. Here we'll look for the "errors" boolean in the response
+        payload, which is true if any failures occurred. We'll then scan the
+        "items" list to diagnose the specific failures in the log.
+
+        FIXME: What do we do about error recovery here? At this point we may
+        have "published" some subset of the Elasticsearch documents but not
+        others. Do we try to "unpublish" those failed documents and fail the
+        whole operation? (And what if some of the "unpublish" operations fail
+        as well?)
+
+        NOTE: The Elasticsearch response "item" looks like this:
+
+                {
+                    "update": {
+                        "_index": "index1",
+                        "_type" : "_doc",
+                        "_id": "5",
+                        "status": 404,
+                        "error": {
+                            "type": "document_missing_exception",
+                            "reason": "[_doc][5]: document missing",
+                            "index_uuid": "aAsFqTI0Tc2W0LCWgPNrOA",
+                            "shard": "0",
+                            "index": "index1"
+                        }
+                    }
+                },
+        """
+        status = 200
+        if es_json["errors"]:
+            for i in es_json["items"]:
+                if i["status"] != HTTPStatus.OK:
+                    e = i["error"]
+                    self.logger.error(
+                        "{} ({}: {}) for id {} in index {}",
+                        e["status"],
+                        e["type"],
+                        e["reason"],
+                        i["_id"],
+                        i["_index"],
+                    )
+                    status = HTTPStatus.FAILED_DEPENDENCY
```

Comment:
I struggled with this choice; anyone have a better idea? It's not really an "internal server error", but it's not a client error either, and Webb's convinced me we shouldn't be using GATEWAY errors ... that doesn't leave a lot of space to maneuver.

Reply:
A comment that I saw suggested that 4XX errors represent problems which the client must deal with, while 5XX errors represent server-side issues...which suggests that `FAILED_DEPENDENCY` isn't appropriate, either (although, otherwise, I kinda like that one).

Nevertheless, I'm still convinced that the server is not acting as a gateway or proxy; rather that, in this instance, it's the client of another service, so it's not appropriate to return either 502 (or 504).

So, that leaves me back at 500...although, we _could_ provide an explanatory message other than "internal server error"....

Reply:
We had generally agreed to obscure internal server error messages; but in this case the details of this case make me a bit nervous, so I'm inclined to agree. I'm adding a specific reference to a failed ES publish operation, and recommending reviewing the server log for details.